datafusion/optimizer/src/subquery_filter_to_join.rs

```diff
@@ -108,13 +107,17 @@ impl OptimizerRule for SubqueryFilterToJoin {
                             };
 
                             let right_key = right_schema.field(0).qualified_column();
-                            let left_key = match *expr.clone() {
-                                Expr::Column(col) => col,
-                                _ => return Err(DataFusionError::NotImplemented(
-                                    "Filtering by expression not implemented for InSubquery"
-                                        .to_string(),
-                                )),
-                            };
+                            let left_key = *expr.clone();
+                            let (on, filter) =
+                                // When left is a constant expression, like 1, 
+                                // the join predicate will be `1 = right_key`, it is better to add it to filter.
+                                if left_key.to_columns()?.is_empty() {
+                                    let equi_expr =
+                                        Expr::eq(*expr.clone(), Expr::Column(right_key));
+                                    (vec![], Some(equi_expr))
+                                } else {
```

Comment:
IMHO, I think it's a good improvement but maybe isn't suitable to put in this rule.🤔
It could be better to handle separate/handle `on/joinfilter` in a new rule(I also contain more improvement).

Just a thought, I think we also can add a TODO, and handle them as a future ticket.

Reply:
> It could be better to handle separate/handle on/joinfilter in a new rule(we also can add more improvement in it).

Agree.

There is another pr #4711 which will separate equi and non-equi predicate from filter.
We can add the `left-key = right-key` to join-filter in this rule, and let the other rule do the separating works.